### PR TITLE
Fix stack overflow with large number of item includes for the same item list

### PR DESCRIPTION
**Pull request by dsplaisted**
Use loops and explicit stacks to traverse the linked list of previous operations instead of recursion.

### CODE REVIEW

File: src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs [hunk 1]
```diff
@@ -147,34 +147,64 @@ public LazyItemList(LazyItemList previous, LazyItemOperation operation)
...
-                }
-                else
-                {
-                    var globsToIgnoreForPreviousOperations = globsToIgnore;
+                //  TODO: Check for cached results
 
+                return GetitemsImplementation(this, globsToIgnore);
+            }
+
+            static ImmutableList<ItemData>.Builder GetitemsImplementation(LazyItemList lazyItemList, ImmutableHashSet<string> globsToIgnore)
```
Comment (rainersigwald) [inline src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs]: Nit capitalize the `i` in `items`.

**Comment by rainersigwald** [comment]
I continue to find this implementation confusing, but this is an improvement over the old version. :shipit:

**Comment by AndyGerlicher** [comment]
:shipit:

**Comment by rainersigwald** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by cdmihai** [comment]
:shipit:

**Comment by cdmihai** [comment]
Thanks!

**Pull request merged**